wire/messages.go

```diff
@@ -0,0 +1,175 @@
+package wire
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/routing"
+
+	logging "github.com/ipfs/go-log"
+	record "github.com/libp2p/go-libp2p-record"
+	recpb "github.com/libp2p/go-libp2p-record/pb"
+	"github.com/multiformats/go-multihash"
+
+	"github.com/libp2p/go-libp2p-kad-dht/internal"
+	pb "github.com/libp2p/go-libp2p-kad-dht/pb"
+)
+
+var logger = logging.Logger("dht")
+
+// ProtocolMessenger can be used for sending DHT messages to peers and processing their responses.
+// This decouples the wire protocol format from both the DHT protocol implementation and from the implementation of the
+// routing.Routing interface.
+//
+// TODO: This is still strongly coupled with the existing implementation of what happens when a peer actually sends a
+// message on the wire (e.g. reusing streams, reusing connections, metrics tracking, etc.).
+type ProtocolMessenger struct {
+	m         MessageSender
+	validator record.Validator
+}
+
+type ProtocolMessengerOption func(*ProtocolMessenger) error
+
+func WithValidator(validator record.Validator) ProtocolMessengerOption {
+	return func(messenger *ProtocolMessenger) error {
+		messenger.validator = validator
+		return nil
+	}
+}
+
+// NewProtocolMessenger creates a new ProtocolMessenger that is used for sending DHT messages to peers and processing
+// their responses.
+func NewProtocolMessenger(msgSender MessageSender, opts ...ProtocolMessengerOption) (*ProtocolMessenger, error) {
+	pm := &ProtocolMessenger{
+		m: msgSender,
+	}
+
+	for _, o := range opts {
+		if err := o(pm); err != nil {
+			return nil, err
+		}
+	}
+
+	return pm, nil
+}
+
+// MessageSender handles sending wire protocol messages to a given peer
+type MessageSender interface {
+	// SendRequest sends a peer a message and waits for its response
+	SendRequest(ctx context.Context, p peer.ID, pmes *pb.Message) (*pb.Message, error)
+	// SendMessage sends a peer a message without waiting on a response
+	SendMessage(ctx context.Context, p peer.ID, pmes *pb.Message) error
```

Comment:
Should these be condensed into a single function at the interface level, or is two reasonable here?

These names were copied from our internal functions, so open to name suggestions here too.

Reply:
two seems reasonable to me